File: contrib/pam_nauthilus/flow.go
package main

import (
	"context"
	"crypto/subtle"
	"crypto/tls"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"strings"
	"time"
)

var (
	// ErrTimeout signals that the device authorization expired or exceeded the timeout.
	ErrTimeout = errors.New("authentication timed out")
	// ErrAccessDenied signals that the user denied the authorization request.
	ErrAccessDenied = errors.New("access denied")
	// ErrUserMismatch signals that the userinfo claim does not match the PAM username.
	ErrUserMismatch = errors.New("user claim does not match PAM user")
)

// Clock abstracts time for testable polling.
type Clock interface {
	Now() time.Time
	Sleep(time.Duration)
}

// systemClock is the production clock implementation.
type systemClock struct{}

// Now returns the current time.
func (systemClock) Now() time.Time {
	return time.Now()
}

// Sleep pauses execution for the given duration.
func (systemClock) Sleep(d time.Duration) {
	time.Sleep(d)
}

// DeviceFlow coordinates the RFC 8628 device authorization flow.
type DeviceFlow struct {
	client         *http.Client
	endpoints      EndpointSet
	clientID       string
	clientSecret   string
	scope          string
	userClaim      string
	requestTimeout time.Duration
	clock          Clock
}

// DeviceAuthorization contains the response values from the device endpoint.
type DeviceAuthorization struct {
	DeviceCode              string
	UserCode                string
	VerificationURI         string
	VerificationURIComplete string
	Interval                time.Duration
	ExpiresIn               time.Duration
}

// TokenResponse contains the access token data returned by the token endpoint.
type TokenResponse struct {
	AccessToken string
	TokenType   string
	ExpiresIn   time.Duration
}

// deviceResponse mirrors the JSON payload from the device endpoint.
type deviceResponse struct {
	DeviceCode              string `json:"device_code"`
	UserCode                string `json:"user_code"`
	VerificationURI         string `json:"verification_uri"`
	VerificationURIComplete string `json:"verification_uri_complete"`
	ExpiresIn               int    `json:"expires_in"`
	Interval                int    `json:"interval"`
	Error                   string `json:"error"`
	ErrorDescription        string `json:"error_description"`
}

// tokenResponse mirrors the JSON payload from the token endpoint.
type tokenResponse struct {
	AccessToken      string `json:"access_token"`
	TokenType        string `json:"token_type"`
	ExpiresIn        int    `json:"expires_in"`
	Error            string `json:"error"`
	ErrorDescription string `json:"error_description"`
}

// NewDeviceFlow builds a DeviceFlow with validated endpoints and secure defaults.
func NewDeviceFlow(settings Settings, client *http.Client, clock Clock) (*DeviceFlow, error) {
	endpoints, err := settings.Endpoints()
	if err != nil {
		return nil, err
	}

	if client == nil {
		client, err = newHTTPClient(settings)
		if err != nil {
			return nil, err
		}
	}

	if clock == nil {
		clock = systemClock{}
	}

	return &DeviceFlow{
		client:         client,
		endpoints:      endpoints,
		clientID:       settings.ClientID,
		clientSecret:   settings.ClientSecret,
		scope:          settings.Scope,
		userClaim:      settings.UserClaim,
		requestTimeout: settings.RequestTimeout,
		clock:          clock,
	}, nil
}

// newHTTPClient configures a TLS-hardened HTTP client for IdP calls.
func newHTTPClient(settings Settings) (*http.Client, error) {
	pool, err := loadCertPool(settings.CAFile)
	if err != nil {
		return nil, err
	}

	tlsConfig := &tls.Config{
		MinVersion: tls.VersionTLS12,
		RootCAs:    pool,
	}

	if settings.TLSServerName != "" {
		tlsConfig.ServerName = settings.TLSServerName
	}

	transport := &http.Transport{
		Proxy:               http.ProxyFromEnvironment,
		TLSClientConfig:     tlsConfig,
		ForceAttemptHTTP2:   true,
		MaxIdleConns:        10,
		IdleConnTimeout:     30 * time.Second,
		TLSHandshakeTimeout: 10 * time.Second,
	}

	return &http.Client{
		Transport: transport,
	}, nil
}

// StartDeviceAuthorization requests a device code and user code from the IdP.
func (f *DeviceFlow) StartDeviceAuthorization(ctx context.Context) (DeviceAuthorization, error) {
	form := url.Values{}
	form.Set("client_id", f.clientID)
	form.Set("scope", f.scope)

	body, status, err := f.postForm(ctx, f.endpoints.Device, form, false)
	if err != nil {
		return DeviceAuthorization{}, err
	}

	var resp deviceResponse
	if err := json.Unmarshal(body, &resp); err != nil {
		return DeviceAuthorization{}, fmt.Errorf("decode device response: %w", err)
	}

	if resp.Error != "" {
		return DeviceAuthorization{}, fmt.Errorf("device authorization error: %s", resp.Error)
	}

	if status != http.StatusOK {
		return DeviceAuthorization{}, fmt.Errorf("device authorization failed: status %d", status)
	}

	if resp.DeviceCode == "" || resp.UserCode == "" || resp.VerificationURI == "" {
		return DeviceAuthorization{}, errors.New("device authorization response missing required fields")
	}

	interval := time.Duration(resp.Interval) * time.Second
	if interval <= 0 {
		interval = 5 * time.Second
	}

	return DeviceAuthorization{
		DeviceCode:              resp.DeviceCode,
		UserCode:                resp.UserCode,
		VerificationURI:         resp.VerificationURI,
		VerificationURIComplete: resp.VerificationURIComplete,
		Interval:                interval,
		ExpiresIn:               time.Duration(resp.ExpiresIn) * time.Second,
	}, nil
}

// PollToken polls the token endpoint until authorization completes or times out.
func (f *DeviceFlow) PollToken(ctx context.Context, deviceCode string, interval time.Duration, deadline time.Time) (TokenResponse, error) {
	if deviceCode == "" {
		return TokenResponse{}, errors.New("device code is required")
	}

	if interval <= 0 {
		interval = 5 * time.Second
	}

	for {
		if !deadline.IsZero() && f.clock.Now().After(deadline) {
			return TokenResponse{}, ErrTimeout
		}

		if err := ctx.Err(); err != nil {
			if errors.Is(err, context.DeadlineExceeded) {
				return TokenResponse{}, ErrTimeout
			}

			return TokenResponse{}, err
		}

		form := url.Values{}
		form.Set("grant_type", "urn:ietf:params:oauth:grant-type:device_code")
		form.Set("device_code", deviceCode)

		body, status, err := f.postForm(ctx, f.endpoints.Token, form, true)
		if err != nil {
			return TokenResponse{}, err
		}

		var resp tokenResponse
		if err := json.Unmarshal(body, &resp); err != nil {
			return TokenResponse{}, fmt.Errorf("decode token response: %w", err)
		}

		if resp.Error != "" {
			switch resp.Error {
			case "authorization_pending":
				f.clock.Sleep(interval)
				continue
			case "slow_down":
				interval += 5 * time.Second
				f.clock.Sleep(interval)
				continue
			case "access_denied":
				return TokenResponse{}, ErrAccessDenied
			case "expired_token":
				return TokenResponse{}, ErrTimeout
			default:
				return TokenResponse{}, fmt.Errorf("token error: %s", resp.Error)
			}
		}

		if status != http.StatusOK {
			return TokenResponse{}, fmt.Errorf("token request failed: status %d", status)
		}

		if resp.AccessToken == "" {
			return TokenResponse{}, errors.New("token response missing access_token")
		}

		if resp.TokenType != "" && !strings.EqualFold(resp.TokenType, "bearer") {
			return TokenResponse{}, fmt.Errorf("unsupported token type: %s", resp.TokenType)
		}

		return TokenResponse{
			AccessToken: resp.AccessToken,
			TokenType:   resp.TokenType,
			ExpiresIn:   time.Duration(resp.ExpiresIn) * time.Second,
		}, nil
	}
}

// FetchUserInfo queries the userinfo endpoint and returns the claims map.
func (f *DeviceFlow) FetchUserInfo(ctx context.Context, accessToken string) (map[string]any, error) {
	if accessToken == "" {
		return nil, errors.New("access token is required")
	}

	reqCtx, cancel := context.WithTimeout(ctx, f.requestTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodGet, f.endpoints.UserInfo, nil)
	if err != nil {
		return nil, err
	}

	req.Header.Set("Authorization", "Bearer "+accessToken)
	req.Header.Set("Accept", "application/json")

	resp, err := f.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("userinfo request failed: status %d", resp.StatusCode)
	}

	var claims map[string]any
	if err := json.Unmarshal(body, &claims); err != nil {
		return nil, fmt.Errorf("decode userinfo response: %w", err)
	}

	return claims, nil
}

// VerifyUser compares the configured claim with the PAM username using a constant-time check.
func (f *DeviceFlow) VerifyUser(claims map[string]any, username string) error {
	claimRaw, ok := claims[f.userClaim]
	if !ok {
		return fmt.Errorf("claim %s not found", f.userClaim)
	}

	claimValue, ok := claimRaw.(string)
	if !ok {
		return fmt.Errorf("claim %s is not a string", f.userClaim)
	}

	if subtle.ConstantTimeCompare([]byte(claimValue), []byte(username)) != 1 {
		return ErrUserMismatch
	}

	return nil
}

// postForm sends a form-encoded request to the IdP and returns body and status.
func (f *DeviceFlow) postForm(ctx context.Context, endpoint string, form url.Values, useBasicAuth bool) ([]byte, int, error) {
	reqCtx, cancel := context.WithTimeout(ctx, f.requestTimeout)
	defer cancel()

	req, err := http.NewRequestWithContext(reqCtx, http.MethodPost, endpoint, strings.NewReader(form.Encode()))
	if err != nil {
		return nil, 0, err
	}

	req.Header.Set("Content-Type", "application/x-www-form-urlencoded")
	req.Header.Set("Accept", "application/json")

	if useBasicAuth {
		req.SetBasicAuth(f.clientID, f.clientSecret)
	}

	resp, err := f.client.Do(req)
	if err != nil {
		return nil, 0, err
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		return nil, resp.StatusCode, err
	}

	return body, resp.StatusCode, nil
}

File: contrib/pam_nauthilus/flow.c
/*
 * flow.c – RFC 8628 Device Authorization Flow implementation.
 */

#include "flow.h"

#include <cjson/cJSON.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h>
#include <time.h>
#include <unistd.h>

/* constant_time_compare returns 0 if a and b are equal, -1 otherwise. */
static int constant_time_compare(const char *a, const char *b)
{
    size_t alen = strlen(a);
    size_t blen = strlen(b);
    size_t maxlen = (alen > blen) ? alen : blen;

    volatile unsigned char result = (alen != blen) ? 1 : 0;

    for (size_t i = 0; i < maxlen; i++) {
        unsigned char ca = (i < alen) ? (unsigned char)a[i] : 0;
        unsigned char cb = (i < blen) ? (unsigned char)b[i] : 0;

        result |= ca ^ cb;
    }

    return (result == 0) ? 0 : -1;
}

/* build_form_data URL-encodes key=value pairs into a static buffer. */
static int build_form_data(char *dst, size_t dstlen,
                           const char *pairs[][2], int count)
{
    dst[0] = '\0';
    size_t pos = 0;

    for (int i = 0; i < count; i++) {
        /* libcurl's curl_easy_escape could be used here, but for simple
         * ASCII values (client_id, scope, device_code) plain concatenation
         * is safe and avoids needing a CURL handle in this function. */
        int n;

        if (pos > 0) {
            n = snprintf(dst + pos, dstlen - pos, "&%s=%s", pairs[i][0], pairs[i][1]);
        } else {
            n = snprintf(dst + pos, dstlen - pos, "%s=%s", pairs[i][0], pairs[i][1]);
        }

        if (n < 0 || (size_t)n >= dstlen - pos) {
            return -1;
        }

        pos += (size_t)n;
    }

    return 0;
}

void token_response_free(token_response *tok)
{
    if (tok == NULL) {
        return;
    }

    free(tok->access_token);
    tok->access_token = NULL;
}

int device_flow_init(device_flow *df, const pam_settings *s,
                     http_client *hc, char *errbuf, size_t errlen)
{
    memset(df, 0, sizeof(*df));

    df->hc = hc;
    df->timeout_sec = s->timeout_sec;
    df->request_timeout_sec = s->request_timeout_sec;

    snprintf(df->client_id, sizeof(df->client_id), "%s", s->client_id);
    snprintf(df->client_secret, sizeof(df->client_secret), "%s", s->client_secret);
    snprintf(df->scope, sizeof(df->scope), "%s", s->scope);
    snprintf(df->user_claim, sizeof(df->user_claim), "%s", s->user_claim);

    return settings_resolve_endpoints(s, &df->endpoints, errbuf, errlen);
}

int device_flow_start_auth(device_flow *df, device_auth *auth,
                           char *errbuf, size_t errlen)
{
    memset(auth, 0, sizeof(*auth));

    char form[2048];
    const char *pairs[][2] = {
        {"client_id", df->client_id},
        {"scope",     df->scope},
    };

    if (build_form_data(form, sizeof(form), pairs, 2) != 0) {
        snprintf(errbuf, errlen, "form data too long");

        return FLOW_ERR_INTERNAL;
    }

    http_buffer buf;

    http_buffer_init(&buf);

    int status = http_post_form(df->hc, df->endpoints.device, form,
                                df->client_id, df->client_secret,
                                0, &buf);

    if (status < 0 || buf.data == NULL) {
        http_buffer_free(&buf);
        snprintf(errbuf, errlen, "device authorization request failed");

        return FLOW_ERR_INTERNAL;
    }

    cJSON *json = cJSON_Parse(buf.data);

    http_buffer_free(&buf);

    if (json == NULL) {
        snprintf(errbuf, errlen, "failed to parse device response JSON");

        return FLOW_ERR_INTERNAL;
    }

    /* Check for error response. */
    cJSON *jerr = cJSON_GetObjectItemCaseSensitive(json, "error");

    if (cJSON_IsString(jerr) && jerr->valuestring[0] != '\0') {
        snprintf(errbuf, errlen, "device authorization error: %s", jerr->valuestring);
        cJSON_Delete(json);

        return FLOW_ERR_INTERNAL;
    }

    if (status != 200) {
        snprintf(errbuf, errlen, "device authorization failed: status %d", status);
        cJSON_Delete(json);

        return FLOW_ERR_INTERNAL;
    }

    /* Extract required fields. */
    cJSON *jdc  = cJSON_GetObjectItemCaseSensitive(json, "device_code");
    cJSON *juc  = cJSON_GetObjectItemCaseSensitive(json, "user_code");
    cJSON *jvu  = cJSON_GetObjectItemCaseSensitive(json, "verification_uri");
    cJSON *jvuc = cJSON_GetObjectItemCaseSensitive(json, "verification_uri_complete");
    cJSON *jei  = cJSON_GetObjectItemCaseSensitive(json, "expires_in");
    cJSON *jiv  = cJSON_GetObjectItemCaseSensitive(json, "interval");

    if (!cJSON_IsString(jdc) || !cJSON_IsString(juc) || !cJSON_IsString(jvu)) {
        snprintf(errbuf, errlen, "device response missing required fields");
        cJSON_Delete(json);

        return FLOW_ERR_INTERNAL;
    }

    snprintf(auth->device_code, sizeof(auth->device_code), "%s", jdc->valuestring);
    snprintf(auth->user_code, sizeof(auth->user_code), "%s", juc->valuestring);
    snprintf(auth->verification_uri, sizeof(auth->verification_uri), "%s", jvu->valuestring);

    if (cJSON_IsString(jvuc)) {
        snprintf(auth->verification_uri_complete, sizeof(auth->verification_uri_complete),
                 "%s", jvuc->valuestring);
    }

    auth->expires_in_sec = cJSON_IsNumber(jei) ? jei->valueint : 0;
    auth->interval_sec = cJSON_IsNumber(jiv) ? jiv->valueint : 5;

    if (auth->interval_sec <= 0) {
        auth->interval_sec = 5;
    }

    cJSON_Delete(json);

    return FLOW_OK;
}

int device_flow_poll_token(device_flow *df, const char *device_code,
                           int interval_sec, int expires_in_sec,
                           token_response *tok,
                           char *errbuf, size_t errlen)
{
    memset(tok, 0, sizeof(*tok));

    if (device_code == NULL || device_code[0] == '\0') {
        snprintf(errbuf, errlen, "device code is required");

        return FLOW_ERR_INTERNAL;
    }

    if (interval_sec <= 0) {
        interval_sec = 5;
    }

    time_t start = time(NULL);
    time_t deadline = start + df->timeout_sec;

    /* Use device expiry if shorter. */
    if (expires_in_sec > 0) {
        time_t dev_deadline = start + expires_in_sec;

        if (dev_deadline < deadline) {
            deadline = dev_deadline;
        }
    }

    char form[2048];
    const char *pairs[][2] = {
        {"grant_type",  "urn:ietf:params:oauth:grant-type:device_code"},
        {"device_code", device_code},
    };

    if (build_form_data(form, sizeof(form), pairs, 2) != 0) {
        snprintf(errbuf, errlen, "form data too long");

        return FLOW_ERR_INTERNAL;
    }

    for (;;) {
        if (time(NULL) >= deadline) {
            return FLOW_ERR_TIMEOUT;
        }

        http_buffer buf;

        http_buffer_init(&buf);

        int status = http_post_form(df->hc, df->endpoints.token, form,
                                    df->client_id, df->client_secret,
                                    1, &buf);

        if (status < 0 || buf.data == NULL) {
            http_buffer_free(&buf);
            snprintf(errbuf, errlen, "token request failed");

            return FLOW_ERR_INTERNAL;
        }

        cJSON *json = cJSON_Parse(buf.data);

        http_buffer_free(&buf);

        if (json == NULL) {
            snprintf(errbuf, errlen, "failed to parse token response JSON");

            return FLOW_ERR_INTERNAL;
        }

        cJSON *jerr = cJSON_GetObjectItemCaseSensitive(json, "error");

        if (cJSON_IsString(jerr) && jerr->valuestring[0] != '\0') {
            const char *errstr = jerr->valuestring;

            if (strcmp(errstr, "authorization_pending") == 0) {
                cJSON_Delete(json);
                sleep((unsigned int)interval_sec);

                continue;
            }

            if (strcmp(errstr, "slow_down") == 0) {
                interval_sec += 5;
                cJSON_Delete(json);
                sleep((unsigned int)interval_sec);

                continue;
            }

            if (strcmp(errstr, "access_denied") == 0) {
                cJSON_Delete(json);

                return FLOW_ERR_DENIED;
            }

            if (strcmp(errstr, "expired_token") == 0) {
                cJSON_Delete(json);

                return FLOW_ERR_TIMEOUT;
            }

            snprintf(errbuf, errlen, "token error: %s", errstr);
            cJSON_Delete(json);

            return FLOW_ERR_INTERNAL;
        }

        if (status != 200) {
            snprintf(errbuf, errlen, "token request failed: status %d", status);
            cJSON_Delete(json);

            return FLOW_ERR_INTERNAL;
        }

        cJSON *jat = cJSON_GetObjectItemCaseSensitive(json, "access_token");

        if (!cJSON_IsString(jat) || jat->valuestring[0] == '\0') {
            snprintf(errbuf, errlen, "token response missing access_token");
            cJSON_Delete(json);

            return FLOW_ERR_INTERNAL;
        }

        cJSON *jtt = cJSON_GetObjectItemCaseSensitive(json, "token_type");

        if (cJSON_IsString(jtt) && jtt->valuestring[0] != '\0') {
            if (strcasecmp(jtt->valuestring, "bearer") != 0) {
                snprintf(errbuf, errlen, "unsupported token type: %s", jtt->valuestring);
                cJSON_Delete(json);

                return FLOW_ERR_INTERNAL;
            }

            snprintf(tok->token_type, sizeof(tok->token_type), "%s", jtt->valuestring);
        }

        tok->access_token = strdup(jat->valuestring);

        cJSON *jei = cJSON_GetObjectItemCaseSensitive(json, "expires_in");

        tok->expires_in_sec = cJSON_IsNumber(jei) ? jei->valueint : 0;

        cJSON_Delete(json);

        return FLOW_OK;
    }
}

int device_flow_fetch_userinfo(device_flow *df, const char *access_token,
                               char *claim_value, size_t claim_len,
                               char *errbuf, size_t errlen)
{
    claim_value[0] = '\0';

    http_buffer buf;

    http_buffer_init(&buf);

    int status = http_get(df->hc, df->endpoints.userinfo, access_token, &buf);

    if (status < 0 || buf.data == NULL) {
        http_buffer_free(&buf);
        snprintf(errbuf, errlen, "userinfo request failed");

        return FLOW_ERR_INTERNAL;
    }

    if (status != 200) {
        http_buffer_free(&buf);
        snprintf(errbuf, errlen, "userinfo request failed: status %d", status);

        return FLOW_ERR_INTERNAL;
    }

    cJSON *json = cJSON_Parse(buf.data);

    http_buffer_free(&buf);

    if (json == NULL) {
        snprintf(errbuf, errlen, "failed to parse userinfo response JSON");

        return FLOW_ERR_INTERNAL;
    }

    cJSON *jclaim = cJSON_GetObjectItemCaseSensitive(json, df->user_claim);

    if (!cJSON_IsString(jclaim)) {
        snprintf(errbuf, errlen, "claim %s not found or not a string", df->user_claim);
        cJSON_Delete(json);

        return FLOW_ERR_INTERNAL;
    }

    snprintf(claim_value, claim_len, "%s", jclaim->valuestring);

    cJSON_Delete(json);

    return FLOW_OK;
}

int device_flow_verify_user(const char *claim_value, const char *username)
{
    return constant_time_compare(claim_value, username);
}
